tests/nexus/platform/nexus_core.cpp

```diff
@@ -0,0 +1,205 @@
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "nexus_core.hpp"
+#include "nexus_node.hpp"
+
+namespace ot {
+namespace Nexus {
+
+Core *Core::sCore = nullptr;
+
+Core::Core(void)
+    : mNow(0)
+    , mCurNodeId(0)
+    , mPendingAction(false)
+{
+    VerifyOrQuit(sCore == nullptr);
+    sCore = this;
+
+    mNextAlarmTime = mNow.GetDistantFuture();
+}
+
+Node &Core::CreateNode(void)
+{
+    Node *node;
+
+    node = Node::Allocate();
+    VerifyOrQuit(node != nullptr);
+
+    node->GetInstance().SetId(mCurNodeId++);
+
+    mNodes.Push(*node);
+
+    node->GetInstance().AfterInit();
+
+    return *node;
+}
+
+void Core::UpdateNextAlarmTime(const Alarm &aAlarm)
+{
+    if (aAlarm.mScheduled)
+    {
+        mNextAlarmTime = Min(mNextAlarmTime, Max(mNow, aAlarm.mAlarmTime));
+    }
+}
+
+void Core::AdvanceTime(uint32_t aDuration)
+{
+    TimeMilli targetTime = mNow + aDuration;
+
+    while (mPendingAction || (mNextAlarmTime <= targetTime))
+    {
+        mNextAlarmTime = mNow.GetDistantFuture();
+        mPendingAction = false;
+
+        for (Node &node : mNodes)
+        {
+            Process(node);
+            UpdateNextAlarmTime(node.mAlarm);
+        }
+
+        if (!mPendingAction)
+        {
+            mNow = Min(mNextAlarmTime, targetTime);
+        }
+    }
+
+    mNow = targetTime;
+}
+
+void Core::Process(Node &aNode)
+{
+    otTaskletsProcess(&aNode.GetInstance());
+
+    ProcessRadio(aNode);
+
+    if (aNode.mAlarm.ShouldTrigger(mNow))
+    {
+        otPlatAlarmMilliFired(&aNode.GetInstance());
+    }
+}
+
+void Core::ProcessRadio(Node &aNode)
+{
+    Mac::Address dstAddr;
+    uint16_t     dstPanId;
+    bool         ackRequested;
+    AckMode      ackMode = kNoAck;
+
+    VerifyOrExit(aNode.mRadio.mState == Radio::kStateTransmit);
+
+    if (aNode.mRadio.mTxFrame.GetDstAddr(dstAddr) != kErrorNone)
+    {
+        dstAddr.SetNone();
+    }
+
+    if (aNode.mRadio.mTxFrame.GetDstPanId(dstPanId) != kErrorNone)
+    {
+        dstPanId = Mac::kPanIdBroadcast;
+    }
+
+    ackRequested = aNode.mRadio.mTxFrame.GetAckRequest();
+
+    otPlatRadioTxStarted(&aNode.GetInstance(), &aNode.mRadio.mTxFrame);
+
+    for (Node &rxNode : mNodes)
+    {
+        bool matchesDst;
+
+        if ((&rxNode == &aNode) || !rxNode.mRadio.CanReceiveOnChannel(aNode.mRadio.mTxFrame.GetChannel()))
+        {
+            continue;
+        }
+
+        matchesDst = rxNode.mRadio.Matches(dstAddr, dstPanId);
+
+        if (matchesDst || rxNode.mRadio.mPromiscuous)
+        {
+            // `rxNode` should receive this frame.
+
+            Radio::Frame rxFrame(aNode.mRadio.mTxFrame);
```

Comment:
Does this assume radio packet transmission time is zero? If so, should we model it?

Reply:
Yes. The current radio emulation model is very simple.

My plan is to add different, more advanced radio emulation models (transmit timing, CSMA timing, interference/collision, frame drop, etc.) over time in follow-up pull requests.

I will likely add them in a way that allows us to select the radio model to use per test. For some situations where we may want to simulate a larger network and test something quickly, a simpler model may suffice and be much faster.